src/Orleans/Utils/InterlockedExchangeLock.cs

```diff
@@ -0,0 +1,40 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Runtime.CompilerServices;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Orleans.Runtime
+{
+    // About 2x faster than ordinary lock on mostly single threaded locking
+    internal class InterlockedExchangeLock
+    {
+        private const int Locked = 1;
+        private const int Unlocked = 0;
+        private int lockState = Unlocked;
+
+        public bool TryGet()
+        {
+            return Interlocked.Exchange(ref lockState, Locked) != Locked;
+        }
+        
+        public void Get()
+        {
+            if (TryGet())
+                return;
+
+            SpinWait spinWait = new SpinWait();
```

Comment:
Mainly because this provides the best performance for 1 producer - 1 consumer scenarios. For example such construction is used in [ConcurrentQueue](https://github.com/dotnet/corefx/blob/master/src/System.Threading.Tasks.Dataflow/src/Internal/ConcurrentQueue.cs#L567).